src/shortestpaths/johnson.jl

```diff
@@ -0,0 +1,93 @@
+
+"""
+    struct JohnsonState{T, U}
+An [`AbstractPathState`](@ref) designed for Johnson shortest-paths calculations.
+"""
+struct JohnsonState{T<:Real, U<:Integer} <: AbstractPathState
+    dists::Matrix{T}
+    parents::Matrix{U}
+end
+
+@doc_str """
+    johnson_shortest_paths(g, distmx=weights(g); parallel=false)
+
+### Implementation Notes
+Use the [Johnson algorithm](https://en.wikipedia.org/wiki/Johnson%27s_algorithm)
+to compute the shortest paths between all pairs of vertices in graph `g` using an
+optional distance matrix `distmx`.
+If the parameter parallel is set true, dijkstra_shortest_paths will run in parallel.
+Parallel bellman_ford_shortest_paths is currently unavailable
+Return a [`LightGraphs.JohnsonState`](@ref) with relevant
+traversal information.
+Behaviour in case of negative cycle depends on bellman_ford_shortest_paths.
+Throws NegativeCycleError() if a negative cycle is present.
+### Performance
+Complexity: O(|V|*|E|)
+A sparse matrix will be produced using distmx.
+To reduce memory overhead,  
+if edge (a, b) does not exist in g then distmx[a, b] should be set to 0.
+Memory overhead: 2*|V|^2 + |E| + |V|
+### Dependencies from LightGraphs
+bellman_ford_shortest_paths
+parallel_multisource_dijkstra_shortest_paths
+dijkstra_shortest_paths
+"""
+function johnson_shortest_paths(
+    g::AbstractGraph{U},
+    distmx::AbstractMatrix{T} = weights(g);
+    parallel::Bool = false
+) where T<:Real where U<:Integer
+
+    nvg = nv(g)
+    #Change when parallel implementation of Bellman Ford available
+    wt_transform = bellman_ford_shortest_paths(g, vertices(g), distmx).dists
+    #Some matrices are immutable
+    new_distmx = sparse(typeof(distmx) == LightGraphs.DefaultDistance ? g : distmx) 
+    #Transform weights
+    for e in edges(g)
+    	new_distmx[src(e), dst(e)] += wt_transform[src(e)] - wt_transform[dst(e)] 
+    end
+    dists = Matrix{T}(nvg, nvg)
```

Comment:
This initialisation is not required if parallel path is taken. You can just initialise dists = Nothing and reinitialise only for sequential dijkstra loop. Same goes with parents.

Reply:
Thanks I'll do that